CORE-9469 Use the Crypto API to fetch signing keys

nkovacsx opened the pull request:
### Overview

Previously we were using `MemberInfo`s to fetch the notary keys for signing. 
Now we use the crypto API to fetch the signing keys that correspond to the notary service's (composite) key.

### Testing

Added an extra test scenario to `LedgerUniquenessCheckerClientServiceImpl` where we test both simple keys as well.
Re-run current smoke/e2e tests.

nkovacsx committed:
CORE-9469 Use the Crypto API to fetch signing keys

nkovacsx committed:
Merge branch 'release/os/5.0' of https://github.com/corda/corda-runtime-os into nandor/CORE-9469/uniqueness-crypto-api-for-signing

 Conflicts:
	gradle.properties

nkovacsx committed:
Merge branch 'release/os/5.0' of https://github.com/corda/corda-runtime-os into nandor/CORE-9469/uniqueness-crypto-api-for-signing

 Conflicts:
	gradle.properties

relyafi commented on components/uniqueness/uniqueness-checker-client-service-impl/src/main/kotlin/net/corda/uniqueness/client/impl/LedgerUniquenessCheckerClientServiceImpl.kt:
```diff
@@ -140,21 +140,23 @@ class LedgerUniquenessCheckerClientServiceImpl @Activate constructor(
-        // Always use the latest registered notary key. Even if key was rotated, the first element will be the latest.
-        val signingNotaryKey = myInfo.notaryKeys.first()
+        // We select the first key that is not null
+        val selectedSigningKey = signingKeys.values.filterNotNull().first()
```
Lines 147 and 151 appear to be doing repeated work. It might be cleaner to condense all of this into something like:

```suggestion
        val selectedSigningKey = signingService
            .findMySigningKeys(setOf(notaryServiceKey)).values
            .filterNotNull()
            .firstOrNull()
            
        require(selectedSigningKey != null) {
            ...
        }
```

nkovacsx replied:
Good point!

relyafi commented on components/uniqueness/uniqueness-checker-client-service-impl/src/main/kotlin/net/corda/uniqueness/client/impl/LedgerUniquenessCheckerClientServiceImpl.kt:
```diff
@@ -140,21 +140,23 @@ class LedgerUniquenessCheckerClientServiceImpl @Activate constructor(
-            "Could not find notary keys to sign with."
+        // We need to make sure there's at least one key we can sign with
+        require(signingKeys.any { it.value != null }) {
+            "Could not find any keys that belongs to the notary service's public key."
```
```suggestion
            "Could not find any keys associated with the notary service's public key."
```

relyafi commented on components/uniqueness/uniqueness-checker-client-service-impl/src/main/kotlin/net/corda/uniqueness/client/impl/LedgerUniquenessCheckerClientServiceImpl.kt:
```diff
@@ -140,21 +140,23 @@ class LedgerUniquenessCheckerClientServiceImpl @Activate constructor(
-            "The notary key selected for signing is not part of the notary service composite key."
+        // We double check that the selected key is actually part of the notary service key
+        require(notaryServiceKey.containsAny(setOf(selectedSigningKey))) {
+            "The notary key selected for signing is not part of the notary service key."
```
```suggestion
            "The notary key selected for signing is not associated with the notary service key."
```

relyafi commented on components/uniqueness/uniqueness-checker-client-service-impl/src/main/kotlin/net/corda/uniqueness/client/impl/LedgerUniquenessCheckerClientServiceImpl.kt:
```diff
@@ -140,21 +140,23 @@ class LedgerUniquenessCheckerClientServiceImpl @Activate constructor(
-        require(notaryServiceKeys.contains(signingNotaryKey)) {
-            "The notary key selected for signing is not part of the notary service composite key."
+        // We double check that the selected key is actually part of the notary service key
+        require(notaryServiceKey.containsAny(setOf(selectedSigningKey))) {
```
```suggestion
        require(notaryServiceKey.isFulfilledBy(selectedSigningKey)) {
```

We've already selected an exact key, so no need to use the collection API? I've had a look in corda-api and I _think_ this is composite key compatible, but you'll need to double check.

nkovacsx replied:
Hmm good point, thanks!

relyafi commented on components/uniqueness/uniqueness-checker-client-service-impl/src/test/kotlin/net/corda/uniqueness/client/impl/LedgerUniquenessCheckerClientServiceImplTest.kt:
```diff
@@ -58,14 +61,56 @@ class LedgerUniquenessCheckerClientServiceImplTest {
     }
 
+    @Test
+    fun `Signing is successful when the provided notary service key is a simple key and the selected key does not match`() {
```
```suggestion
    fun `Signing is unsuccessful when the provided notary service key is a simple key and the selected key does not match`() {
```

nkovacsx committed:
PR comment fixes

nkovacsx committed:
Fix tests

nkovacsx committed:
vbump

nkovacsx committed:
Merge branch 'release/os/5.0' of https://github.com/corda/corda-runtime-os into nandor/CORE-9469/uniqueness-crypto-api-for-signing

 Conflicts:
	gradle.properties

relyafi reviewed (dismissed):
LGTM

blsemo reviewed (dismissed):
LGTM

nkovacsx committed:
Merge branch 'release/os/5.0' of https://github.com/corda/corda-runtime-os into nandor/CORE-9469/uniqueness-crypto-api-for-signing

 Conflicts:
	gradle.properties

nkovacsx committed:
vbump

nkovacsx committed:
dummy-test

nkovacsx commented:
Removed `isFulfilledBy` as it seems it's not exactly what we want. It checks that the provided key(s) completely fulfils the composite key which is not true in our case. Replaced it with `containsAny` which will be a double check but it can't hurt.

nkovacsx committed:
vbump
